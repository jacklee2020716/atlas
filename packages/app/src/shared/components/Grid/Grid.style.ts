import { StyleFn, makeStyles } from '../../utils'

export type GridStyleProps = {
  minItemWidth?: string | number
  maxItemWidth?: string | number
}
const container: StyleFn = (_, { minItemWidth = '300', maxItemWidth }) => ({
  display: 'grid',
  gridTemplateColumns: `repeat(auto-fit, minmax(${minItemWidth}px, ${maxItemWidth ? `${maxItemWidth}px` : '1fr'}))`,
  gap: '30px',
})

const item: StyleFn = () => ({
  cursor: 'pointer',
  width: '100%',
})

export const useCSS = (props: GridStyleProps) => ({
  container: makeStyles([container])(props),
  item: makeStyles([item])(props),
})
