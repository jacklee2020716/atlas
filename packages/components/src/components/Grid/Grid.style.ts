import { css } from "@emotion/core"

export type GridStyleProps = {
  minItemWidth?: string | number
  maxItemWidth?: string | number
}

export let makeStyles = ({
  minItemWidth = "300",
  maxItemWidth
}: GridStyleProps) => {
  return {
    container: css`
      display: grid;
      grid-template-columns: repeat(auto-fit, minmax(${minItemWidth}px, ${maxItemWidth ? maxItemWidth + "px" : "1fr"}));
      gap: 30px;
    `,
    item: css`
      width: 100%;
      cursor: pointer;
    `,
  }
}
