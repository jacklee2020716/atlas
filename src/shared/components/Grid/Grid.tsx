import React from 'react'
import styled from '@emotion/styled'
import useResizeObserver from 'use-resize-observer'
import { useMediaQuery } from 'react-responsive'
import { spacing } from '../../theme'

const toPx = (n: number | string) => (typeof n === 'number' ? `${n}px` : n)

const LARGE_VIEWPORT_BREAKPOINT = toPx(2000)

type ContainerProps = {
  gap: number | string
  gridTemplateColumns: string
}

const Container = styled.div<ContainerProps>`
  display: grid;
  gap: ${(props) => toPx(props.gap)};
  grid-template-columns: ${(props) => props.gridTemplateColumns};
`

type GridProps = {
  gap?: number | string
  className?: string
  maxColumns?: number
  minWidth?: number | string
  repeat?: 'fit' | 'fill'
  onResize?: (sizes: number[]) => void
}

const Grid: React.FC<GridProps> = ({
  className,
  gap = spacing.xl,
  onResize,
  repeat = 'fit',
  maxColumns = 6,
  minWidth = 300,
  ...props
}) => {
  const { ref: gridRef } = useResizeObserver<HTMLDivElement>({
    onResize: () => {
      if (onResize && gridRef.current) {
        const computedStyles = window.getComputedStyle(gridRef.current)
        const columnSizes = computedStyles.gridTemplateColumns.split(' ').map(parseFloat)
        onResize(columnSizes)
      }
    },
  })
  const isLargeViewport = useMediaQuery({ query: `(min-width: ${LARGE_VIEWPORT_BREAKPOINT})` })

  const gridTemplateColumns = isLargeViewport
    ? `repeat(${maxColumns}, 1fr)`
    : `repeat(auto-${repeat}, minmax(min(${toPx(minWidth)}, 100%), 1fr))`

  return (
    <Container {...props} className={className} ref={gridRef} gap={gap} gridTemplateColumns={gridTemplateColumns} />
  )
}
export default Grid
