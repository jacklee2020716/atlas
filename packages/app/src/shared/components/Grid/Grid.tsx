import React from 'react'
import { GridStyleProps, useCSS } from './Grid.style'

type SectionProps = {
  items?: React.ReactNode[]
  className?: string
} & GridStyleProps

export default function Grid({ items = [], className = '', ...styleProps }: SectionProps) {
  let styles = useCSS(styleProps)
  return (
    <div css={styles.container} className={className}>
      {items.map((item, index) => (
        <div key={`grid-item-${index}`} css={styles.item}>
          {item}
        </div>
      ))}
    </div>
  )
}
