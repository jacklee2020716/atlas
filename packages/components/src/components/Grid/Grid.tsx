import React from "react"
import { makeStyles, GridStyleProps } from "./Grid.style"

type SectionProps = {
  items?: React.ReactNode[]
  className?: string
} & GridStyleProps

export default function Grid({
  items = [],
  className = "",
  ...styleProps
}: SectionProps) {
  let styles = makeStyles(styleProps)
  return (
    <div css={styles.container} className={className}>
      {items.map((item, index) => <div key={`grid-item-${index}`} css={styles.item}>{item}</div>)}
    </div>
  )
}
