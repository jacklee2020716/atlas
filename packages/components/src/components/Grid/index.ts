import { memo } from "react"
import Grid from "./Grid"
export default memo(Grid)
