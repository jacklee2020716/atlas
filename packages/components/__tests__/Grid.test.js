import React from "react"
import { mount } from "enzyme"
import Grid from "./../src/components/Grid"

describe("Grid component", () => {

  const Item = ({ text }) => <div>{text}</div>

  const component = mount(<Grid
      minItemWidth="250"
      maxItemWidth="600"
      items={[...Array(10).keys()].map((i, index) => <Item key={index} text={`test-${i}`} />)}
    />)

  it("Should render correctly", () => {
    expect(component).toMatchSnapshot()
  })

})