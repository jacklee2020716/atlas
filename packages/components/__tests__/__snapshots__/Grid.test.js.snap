// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Grid component Should render correctly 1`] = `
<Memo(Grid)
  items={
    Array [
      <Item
        text="test-0"
      />,
      <Item
        text="test-1"
      />,
      <Item
        text="test-2"
      />,
      <Item
        text="test-3"
      />,
      <Item
        text="test-4"
      />,
      <Item
        text="test-5"
      />,
      <Item
        text="test-6"
      />,
      <Item
        text="test-7"
      />,
      <Item
        text="test-8"
      />,
      <Item
        text="test-9"
      />,
    ]
  }
  maxItemWidth="600"
  minItemWidth="250"
>
  <div
    className=""
    css={
      Object {
        "map": undefined,
        "name": "1sl9x6x",
        "next": undefined,
        "styles": "
      display: grid;
      grid-template-columns: repeat(auto-fit, minmax(250px, 600px));
      gap: 30px;
    ",
        "toString": [Function],
      }
    }
  >
    <div
      css={
        Object {
          "map": undefined,
          "name": "wuoxs",
          "next": undefined,
          "styles": "
      width: 100%;
      cursor: pointer;
    ",
          "toString": [Function],
        }
      }
      key="grid-item-0"
    >
      <Item
        key="0"
        text="test-0"
      >
        <div>
          test-0
        </div>
      </Item>
    </div>
    <div
      css={
        Object {
          "map": undefined,
          "name": "wuoxs",
          "next": undefined,
          "styles": "
      width: 100%;
      cursor: pointer;
    ",
          "toString": [Function],
        }
      }
      key="grid-item-1"
    >
      <Item
        key="1"
        text="test-1"
      >
        <div>
          test-1
        </div>
      </Item>
    </div>
    <div
      css={
        Object {
          "map": undefined,
          "name": "wuoxs",
          "next": undefined,
          "styles": "
      width: 100%;
      cursor: pointer;
    ",
          "toString": [Function],
        }
      }
      key="grid-item-2"
    >
      <Item
        key="2"
        text="test-2"
      >
        <div>
          test-2
        </div>
      </Item>
    </div>
    <div
      css={
        Object {
          "map": undefined,
          "name": "wuoxs",
          "next": undefined,
          "styles": "
      width: 100%;
      cursor: pointer;
    ",
          "toString": [Function],
        }
      }
      key="grid-item-3"
    >
      <Item
        key="3"
        text="test-3"
      >
        <div>
          test-3
        </div>
      </Item>
    </div>
    <div
      css={
        Object {
          "map": undefined,
          "name": "wuoxs",
          "next": undefined,
          "styles": "
      width: 100%;
      cursor: pointer;
    ",
          "toString": [Function],
        }
      }
      key="grid-item-4"
    >
      <Item
        key="4"
        text="test-4"
      >
        <div>
          test-4
        </div>
      </Item>
    </div>
    <div
      css={
        Object {
          "map": undefined,
          "name": "wuoxs",
          "next": undefined,
          "styles": "
      width: 100%;
      cursor: pointer;
    ",
          "toString": [Function],
        }
      }
      key="grid-item-5"
    >
      <Item
        key="5"
        text="test-5"
      >
        <div>
          test-5
        </div>
      </Item>
    </div>
    <div
      css={
        Object {
          "map": undefined,
          "name": "wuoxs",
          "next": undefined,
          "styles": "
      width: 100%;
      cursor: pointer;
    ",
          "toString": [Function],
        }
      }
      key="grid-item-6"
    >
      <Item
        key="6"
        text="test-6"
      >
        <div>
          test-6
        </div>
      </Item>
    </div>
    <div
      css={
        Object {
          "map": undefined,
          "name": "wuoxs",
          "next": undefined,
          "styles": "
      width: 100%;
      cursor: pointer;
    ",
          "toString": [Function],
        }
      }
      key="grid-item-7"
    >
      <Item
        key="7"
        text="test-7"
      >
        <div>
          test-7
        </div>
      </Item>
    </div>
    <div
      css={
        Object {
          "map": undefined,
          "name": "wuoxs",
          "next": undefined,
          "styles": "
      width: 100%;
      cursor: pointer;
    ",
          "toString": [Function],
        }
      }
      key="grid-item-8"
    >
      <Item
        key="8"
        text="test-8"
      >
        <div>
          test-8
        </div>
      </Item>
    </div>
    <div
      css={
        Object {
          "map": undefined,
          "name": "wuoxs",
          "next": undefined,
          "styles": "
      width: 100%;
      cursor: pointer;
    ",
          "toString": [Function],
        }
      }
      key="grid-item-9"
    >
      <Item
        key="9"
        text="test-9"
      >
        <div>
          test-9
        </div>
      </Item>
    </div>
  </div>
</Memo(Grid)>
`;
